src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+            // For example, if an index contains both an insertion and deletion, we can combine the two
+            // edits into one.
+            // We can also combine edits if we have consecutive edits of certain types:
+            // Delete->Delete, Insert->Insert, Update->Update, Update->Insert, and Update->Delete.
+            // Note for the Update->Insert and Update->Delete cases, any further edits we combine can
+            // only be an Insert or Delete, respectively.
+
+            using var _ = ArrayBuilder<LSP.SemanticTokensEdit>.GetInstance(out var semanticTokensEdits);
+
+            // We sort the edit kinds by index since we need to know what kind of edits surround
+            // a given index in order to potentially combine them into one edit.
+            var editIndexes = indexToEditKinds.Keys.ToArray();
+            Array.Sort(editIndexes);
+
+            for (var i = 0; i < editIndexes.Length; i++)
+            {
+                var initialEditIndex = editIndexes[i];
+                var initialEditKind = indexToEditKinds[initialEditIndex];
+
+                if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert) && initialEditKind.HasFlag(SemanticTokenEditKind.Delete))
+                {
+                    i = AddInsertionDeletionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+                else if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert))
+                {
+                    i = AddInsertionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+                else
+                {
+                    i = AddDeletionEdit(
+                        indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+            }
+
+            return semanticTokensEdits.ToArray();
+
+            static int AddInsertionDeletionEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndexes,
+                int i,
+                int initialEditIndex)
+            {
+                var deleteCount = 5;
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(newGroupedSemanticTokens[initialEditIndex].ToArray());
+
+                // An "update" (i.e. a dual insertion/deletion) can be combined with an update,
+                // deletion, or insertion that directly follows it. If combined with an insertion
+                // or deletion, only that type is allowed in the combined edit afterwards.
+                var editKind = SemanticTokenEditKind.None;
```

Comment:
starting this with 'none' feels weird. don't you know what you started with at 'i' (i.e. a delete/insert/update)?